### PR TITLE
WIP support for Scan

### LINKED ISSUE #105: Add support for `Scan`
Fold : foldl :: Scan : mapAccumL 
```haskell
data Scan a b  = forall x. Scan (x -> a -> (x, b)) x
```
`prescan` and `postscan` along aren't good enough because there's no way to recover the identity `Scan`:
```haskell
id :: Scan a a
id = Scan (\()-> \a -> ((), a)) () 
```

### PR DESCRIPTION
**Pull request by Ericson2314**
Fixes #105

### CODE REVIEW

File: src/Control/Scanl.hs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import Control.Category
+import Control.Foldl hiding (premap, prescan, postscan)
+import Control.Foldl.Internal (Pair(..))
+import Data.Profunctor
+import Prelude hiding ((.), id)
+
+{-| Efficient representation of a left map-with-accumulator that preserves the
+    scan's step function and initial accumulator.
+
+    This allows the 'Applicative' instance to assemble derived folds that
```
Comment (Gabriel439) [inline src/Control/Scanl.hs]: `s/derived folds/derived scans/`

File: src/Control/Scanl.hs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+instance Arrow Scan where
+    arr f = Scan (\()-> \a -> ((), f a)) ()
+    {-# INLINE arr #-}
+    first  (Scan step begin) = Scan (\x (a,b) -> fmap (,b) $ step x a) begin
+    {-# INLINE first #-}
+    second (Scan step begin) = Scan (\x (b,a) -> fmap (b,) $ step x a) begin
+    {-# INLINE second #-}
+
+
+#if MIN_VERSION_base(4,8,0)
```
Comment (Gabriel439) [inline src/Control/Scanl.hs]: What is the reason for this condition on the version of `base`?

Comment (Ericson2314) [reply src/Control/Scanl.hs]: I cargo culted from Fold

**Comment by Ericson2314** [comment]
I figure most scans can be created from folds, so we'll need less in this module? But we could also blanket reexport them.

**Comment by Ericson2314** [comment]
Figured out the reason for the CPP and trying to fix it.

**Comment by Gabriel439** [comment]
@Ericson2314: I think it's fine if the suite of `Scan` utilities is sparse for now or requires wrapping `Fold`s into `Scan`s.  Just having the type and the instances is already a huge improvement

**Comment by Ericson2314** [comment]
Oh pre AMP haskell...

**Comment by Ericson2314** [comment]
Looking at the failures, oh pre AMP Haskell...

**Comment by Gabriel439** [comment]
@Ericson2314: I'm fine dropping support for those versions.  Let me update the Travis Script to omit them

**Comment by Gabriel439** [comment]
@Ericson2314: Alright, if you merge `master` into your branch then it should turn green

**Comment by Ericson2314** [comment]
@Gabriel439 just rebased onto master after my other PR was merged, but this was green and should turn green again. Anything else that should be done here?

**Comment by Gabriel439** [comment]
This is perfect!  Thanks a lot for all your work on this 🙂

**Pull request merged**